includes/class-wc-product-variable.php

```diff
@@ -285,10 +292,13 @@ public function get_virtual( $context = 'view' ) {
 	 * @return array
 	 */
 	public function get_available_variations() {
-
+		global $wp_filter;
 		$variation_ids        = $this->get_children();
 		$available_variations = array();
 
+		// @var bool Whether available variations are cacheable or not.
+		$available_variations_cacheable = empty( $wp_filter['woocommerce_show_variation_price'] ) && empty( $wp_filter['woocommerce_available_variation'] );
```

Comment:
I think you could use `has_filter` function here, unless it's considerably slower.

I believe we need to check for filters called in the sub-functions as well, which, when looking at it now, could be a bit of a problem, as it seems to me there are 91 distinct filters called 253 times in total... what's your take?
```
Called filter 'alloptions'
Called filter 'block_parser_class'
Called filter 'content_url'
Called filter 'get_attached_file'
Called filter 'get_post_metadata'
Called filter 'get_post_status'
Called filter 'gettext'
Called filter 'image_downsize'
Called filter 'option_siteurl'
Called filter 'option_stylesheet'
Called filter 'option_template'
Called filter 'option_theme_mods_storefront'
Called filter 'option__transient_product-transient-version'
Called filter 'option_use_smilies'
Called filter 'option_woocommerce_calc_taxes'
Called filter 'option_woocommerce_manage_stock'
Called filter 'option_woocommerce_prices_include_tax'
Called filter 'option_woocommerce_tax_display_shop'
Called filter 'pre_option_siteurl'
Called filter 'pre_option_stylesheet'
Called filter 'pre_option_template'
Called filter 'pre_option_theme_mods_storefront'
Called filter 'pre_option__transient_product-transient-version'
Called filter 'pre_option_use_smilies'
Called filter 'pre_option_woocommerce_calc_taxes'
Called filter 'pre_option_woocommerce_manage_stock'
Called filter 'pre_option_woocommerce_prices_include_tax'
Called filter 'pre_option_woocommerce_tax_display_shop'
Called filter 'pre_transient_product-transient-version'
Called filter 'set_url_scheme'
Called filter 'stylesheet'
Called filter 'stylesheet_directory_uri'
Called filter 'template'
Called filter 'template_directory_uri'
Called filter 'theme_mod_sp_product_gallery_layout'
Called filter 'theme_root_uri'
Called filter 'transient_product-transient-version'
Called filter 'wc_tax_enabled'
Called filter 'woocommerce_available_variation'
Called filter 'woocommerce_format_content'
Called filter 'woocommerce_format_dimensions'
Called filter 'woocommerce_format_localized_decimal'
Called filter 'woocommerce_format_weight'
Called filter 'woocommerce_gallery_full_size'
Called filter 'woocommerce_gallery_image_size'
Called filter 'woocommerce_gallery_thumbnail_size'
Called filter 'woocommerce_get_availability'
Called filter 'woocommerce_get_availability_class'
Called filter 'woocommerce_get_availability_text'
Called filter 'woocommerce_get_price_excluding_tax'
Called filter 'woocommerce_get_stock_html'
Called filter 'woocommerce_get_variation_prices_hash'
Called filter 'woocommerce_get_variation_regular_price'
Called filter 'woocommerce_get_variation_sale_price'
Called filter 'woocommerce_is_downloadable'
Called filter 'woocommerce_is_purchasable'
Called filter 'woocommerce_is_sold_individually'
Called filter 'woocommerce_is_virtual'
Called filter 'woocommerce_prices_include_tax'
Called filter 'woocommerce_product_backorders_allowed'
Called filter 'woocommerce_product_is_in_stock'
Called filter 'woocommerce_product_is_taxable'
Called filter 'woocommerce_product_thumbnails_large_size'
Called filter 'woocommerce_product_variation_get_attributes'
Called filter 'woocommerce_product_variation_get_backorders'
Called filter 'woocommerce_product_variation_get_description'
Called filter 'woocommerce_product_variation_get_downloadable'
Called filter 'woocommerce_product_variation_get_height'
Called filter 'woocommerce_product_variation_get_image_id'
Called filter 'woocommerce_product_variation_get_length'
Called filter 'woocommerce_product_variation_get_manage_stock'
Called filter 'woocommerce_product_variation_get_parent_id'
Called filter 'woocommerce_product_variation_get_price'
Called filter 'woocommerce_product_variation_get_regular_price'
Called filter 'woocommerce_product_variation_get_sku'
Called filter 'woocommerce_product_variation_get_sold_individually'
Called filter 'woocommerce_product_variation_get_status'
Called filter 'woocommerce_product_variation_get_stock_status'
Called filter 'woocommerce_product_variation_get_tax_status'
Called filter 'woocommerce_product_variation_get_virtual'
Called filter 'woocommerce_product_variation_get_weight'
Called filter 'woocommerce_product_variation_get_width'
Called filter 'woocommerce_resize_images'
Called filter 'woocommerce_short_description'
Called filter 'woocommerce_show_variation_price'
Called filter 'woocommerce_thumbnail_size'
Called filter 'woocommerce_variation_is_active'
Called filter 'woocommerce_variation_is_purchasable'
Called filter 'woocommerce_variation_is_visible'
Called filter 'wp_get_attachment_image_src'
Called filter 'wp_get_attachment_metadata'
```

Reply:
hi @peterfabian, good point. IMHO this is not a good way to go because of maintainability reasons. Also, this PR is relevant for variant-intensive sites, and in such a case is most probable that some of this filters are used.

And taking this consideration, from the initial proposal to now, we are going close and close to only cache `wc_get_product`. This function is already cached in new WooCommerce versions and any other processing that can be cached, for sure, can vary if has filters.

So right now my best criteria says that don't make sense to keep working on this. So unless anyone in the Woo team considers this work specially relevant or we can assume cache some filter processing I would suggest to close this PR because don't makes sense right now since `wc_get_product` is already cached.

Thank you all for all your work here!

Reply:
Thanks for your effort, @juliquiron, this is a tricky one, especially if you need to account for different possible use cases. I agree that with this list of filters, it does not make sense. 

Additionally, when thinking about it now, even checking all those filters would not help us anyway, if we were to be super strict. Any code can theoretically add hooks at any time and remove them again, so...
1. check for hooks? nothing hooked, ok
2. run `get_available_variation`
  a) run hook 1 that adds function _f_ to hook 2
  b) run hook 2 with a new function
  c) run hook 3 that unhooks function _f_ from hook 2
3. cache result of `get_available_variation`

following call:
1. check for hooks? nothing hooked, ok
2. returned cached result

However, function _f_ still needs to run each time (as described in the [comment above](https://github.com/woocommerce/woocommerce/pull/23148#issuecomment-509625459)), so even though there were no hooks in the first run, we'll be getting incorrect results. 

Highly improbable, but perhaps possible...